autoload/lsp/internal/linked_editing_range.vim

```diff
@@ -0,0 +1,136 @@
+
+let s:TEXT_MARK_NAMESPACE = 'lsp#internal#linked_editing_range'
+
+let s:state = {}
+let s:state['bufnr'] = -1
+let s:state['changenr'] = -1
+let s:state['changedtick'] = -1
+
+function! lsp#internal#linked_editing_range#_enable() abort
+    if !s:TextEdit.is_text_mark_preserved()
+        return
+    endif
+
+    if !g:lsp_linked_editing_range_enabled | return | endif
+    let s:Dispose = lsp#callbag#merge(
+    \     lsp#callbag#pipe(
+    \         lsp#callbag#fromEvent(['InsertEnter']),
+    \         lsp#callbag#flatMap({ -> s:request(v:false) }),
+    \         lsp#callbag#subscribe({
+    \           'next': { x -> s:prepare(x) }
+    \         })
+    \     ),
+    \     lsp#callbag#pipe(
+    \         lsp#callbag#fromEvent(['InsertLeave']),
+    \         lsp#callbag#subscribe({ -> s:clear() })
+    \     ),
+    \     lsp#callbag#pipe(
+    \         lsp#callbag#fromEvent(['TextChanged', 'TextChangedI', 'TextChangedP']),
+    \         lsp#callbag#delay(0),
+    \         lsp#callbag#subscribe({ -> s:sync() })
+    \     ),
+    \ )
+endfunction
+
+function! lsp#internal#linked_editing_range#_disable() abort
+    if exists('s:Dispose')
+        call s:clear()
+        call s:Dispose()
+        unlet s:Dispose
+    endif
+endfunction
+
+function! s:request(sync) abort
+    let l:server = lsp#get_allowed_servers(&filetype)
+    let l:server = filter(l:server, 'lsp#capabilities#has_linked_editing_range_provider(v:val)')
+    let l:server = get(l:server, 0, v:null)
+    if empty(l:server)
+        return lsp#callbag#empty()
+    endif
+
+    let l:X = lsp#callbag#pipe(
+    \     lsp#request(l:server, {
+    \         'method': 'textDocument/linkedEditingRange',
+    \         'params': {
+    \             'textDocument': lsp#get_text_document_identifier(),
+    \             'position': lsp#get_position(),
+    \         }
+    \     }),
+    \ )
+    if a:sync
+      return lsp#callbag#of(
+      \     get(
+      \         lsp#callbag#pipe(
+      \             l:X,
+      \             lsp#callbag#toList()
+      \         ).wait({ 'sleep': 1, 'timeout': 200 }),
+      \        0,
+      \        v:null
+      \     )
+      \ )
+    endif
+    return l:X
+endfunction
+
+function! s:prepare(x) abort
+    if empty(get(a:x['response']['result'], 'ranges', {}))
+      return
+    endif
+
+    call s:TextMark.set(bufnr('%'), s:TEXT_MARK_NAMESPACE, map(a:x['response']['result']['ranges'], { _, range -> {
+    \     'range': range,
+    \     'highlight': 'Underlined',
+    \ } }))
+    let s:state['bufnr'] = bufnr('%')
+    let s:state['changenr'] = changenr()
+    let s:state['changedtick'] = b:changedtick
+endfunction
+
+function! s:clear() abort
+    call s:TextMark.clear(bufnr('%'), s:TEXT_MARK_NAMESPACE)
+endfunction
+
+function! s:sync() abort
+    let l:bufnr = bufnr('%')
+    if s:state['bufnr'] != l:bufnr
+        return
+    endif
+    if s:state['changedtick'] == b:changedtick
```

Comment:
ignore `TextChanged` if it caused by `s:sync` itself so I check changedtick in here.